shell/platform/linux/fl_renderer.cc

```diff
@@ -27,12 +48,46 @@ typedef struct {
   // was rendered
   bool had_first_frame;
 
-  GdkGLContext* main_context;
-  GdkGLContext* resource_context;
+  // Shader program.
+  GLuint program;
+
+  // Textures to render.
+  GPtrArray* textures;
 } FlRendererPrivate;
 
 G_DEFINE_TYPE_WITH_PRIVATE(FlRenderer, fl_renderer, G_TYPE_OBJECT)
 
+// Returns the log for the given OpenGL shader.
+static gchar* get_shader_log(GLuint shader) {
+  int log_length;
+  gchar* log;
+
+  glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &log_length);
+
+  log = static_cast<gchar*>(g_malloc(log_length + 1));
+  glGetShaderInfoLog(shader, log_length, nullptr, log);
```

Comment:
Here and a couple other places; should we be calling glGetError to check results on some of these calls. I just grepped the source and admittedly it looks like we almost never do, so perhaps that's a cleanup we can leave for another time. Up to you.

Reply:
If you run with `GDK_DEBUG=gl-debug` then GTK uses the `GL_KHR_debug` extension which reports all GL errors as they occur (including text description). Not sure how valuable the `glGetError` calls are, though they would be valuable in debugging random crashes.